src/coreclr/src/jit/morph.cpp

```diff
@@ -5759,7 +5758,7 @@ GenTree* Compiler::fgMorphStackArgForVarArgs(unsigned lclNum, var_types varType,
         // Create a node representing the local pointing to the base of the args
         GenTree* ptrArg =
             gtNewOperNode(GT_SUB, TYP_I_IMPL, gtNewLclvNode(lvaVarargsBaseOfStkArgs, TYP_I_IMPL),
-                          gtNewIconNode(varDsc->lvStkOffs - codeGen->intRegState.rsCalleeRegArgCount * REGSIZE_BYTES +
+                          gtNewIconNode(varDsc->lvStkOffs - codeGen->intRegState.rsCalleeRegArgCount * REGSIZE_BYTES -
```

Comment:
So ... this seems correct, in that we're subtracting so we need to reverse the sense of `lclOffs` - was this simply incorrect before and went unnoticed?

Reply:
this is x86 vararg only code, I do not think we have many tests for it.

Reply:
>  was this simply incorrect before and went unnoticed?

Looks like. It's sure that fields always have positive offsets so there's no way `+` would have worked because the offset gets subtracted. The common "issue" in this change seems to be "we never had LCL_FLDs before global morph so we keep hitting various code paths that aren't prepared to deal with it".

Well, the importer does generate a few LCL_FLDs but they're so rare and specific that probably don't matter. Either those don't hit these code paths or you need such code contortions to make it happen that it's irrelevant.

> this is x86 vararg only code, I do not think we have many tests for it.

True. But this was actually caught by a failed test - "JIT\Directed\arglist\vararg". I diffed the dumps and noticed the changed offsets. With this fix the offsets are back to normal.

Reply:
but after some drawing, I think that `varDsc->lvStkOffs` and `lclOffs` are both positive and should have the same sign, so the previous version looks correct to me.
```
        GenTree* firstStackVarArg = gtNewLclvNode(lvaVarargsBaseOfStkArgs, TYP_I_IMPL);
        int stackOffset = varDsc->lvStkOffs - codeGen->intRegState.rsCalleeRegArgCount * REGSIZE_BYTES + lclOffs;
        GenTree* stackOffsetNode = gtNewIconNode(stackOffset);
        GenTree* ptrArg = gtNewOperNode(GT_SUB, TYP_I_IMPL, firstStackVarArg, stackOffsetNode);

```

stack | address
-- | --
firstStackVarArg + 0x00 (size 4) | 1008 <- firstStackVarArg
firstStackVarArg + 0x01 | 1007
firstStackVarArg + 0x02 | 1006
firstStackVarArg + 0x03 | 1005
ourLclVar + 0x00 (lclOffs = 6, size = 8) | 1004, so varDsc->lvStkOffs = 4
ourLclVar + 0x01 | 1003
ourLclVar + 0x02 | 1002
ourLclVar + 0x03 | 1001
ourLclVar + 0x04 | 1000
ourLclVar + 0x05 | 999
ourLclVar + 0x06 <- where we really pointing | 998, so the final offset is stackOffset=6 + 4   = 10
ourLclVar + 0x07 | 997

But I am probably wrong, in any case it would be useful to have more locals here (for the const offset at least), and maybe an asserts about offsets signs.

Reply:
I'm not sure I understand the table above. How can "firstStackVarArg + 0x01" at address 1007 if "firstStackVarArg + 0x00" is at 1008? Presumably "address" isn't an absolute memory address but a stack offset?

Ultimately there's no way for a struct field to be at a negative offset. It may be that the argument itself is at a negative offset from the stack pointer but once you compute the base address of the argument the field is at a positive offset.

Reply:
> Presumably "address" isn't an absolute memory address but a stack offset?

Thanks, now I see my mistake, ignore my previous comment.

Reply:
For more clarity here's the original tree in the failing test:
```
               [000016] ------------                 \--*  ADD       int   
               [000012] ------------                    +--*  FIELD     int    a
               [000011] ------------                    |  \--*  ADDR      byref 
               [000010] ------------                    |     \--*  LCL_VAR   struct<NativeVarargTest.TwoIntStruct, 8> V01 arg1         
               [000015] ------------                    \--*  FIELD     int    b
               [000014] ------------                       \--*  ADDR      byref 
               [000013] ------------                          \--*  LCL_VAR   struct<NativeVarargTest.TwoIntStruct, 8> V01 arg1         
```
`LCL_VAR` nodes were first morphed by `fgMorphStackArgForVarArgs`. Of course that `lclOffs` was 0 and the bug didn't manifest. Then `FIELD`s were morphed so we had something like `IND(ADD(ADDR(IND(SUB(LCL_VAR, 16))), 4)` (oh the horror) which folded to `IND(ADD(LCL_VAR, -12))`.

With the `LCL_FLD` produced early we generate `IND(ADD(LCL_VAR, -12))` straight from `fgMorphStackArgForVarArgs`.

Reply:
It turns out that it was possible for an importer generated `LCL_FLD` to hit this path before my change but `LCL_FLD`s generated by the importer seem to always have 0 offset. They are used to reinterpret call returned values.